### ISSUE TITLE
Introduction to Neural Network

### ISSUE THREAD

**Issue by Kulbear** [Kulbear/Schedule#2]
You can find this article and source code at my [GitHub repo](https://github.com/Kulbear/deep-learning-nano-foundation)

## A Deadly Simple Neural Network

If you ever heard of or look at some materials about the neural network, a model looks like the one in below figure should not be a stranger to you.

![An artificial neural network model](https://cloud.githubusercontent.com/assets/14886380/22799690/b0d8c4a2-ef41-11e6-9df9-138ef82b017a.png)

Well, that's a little bit complicated, so what about this one?

![Diagram of a simple neural network](https://cloud.githubusercontent.com/assets/14886380/22798490/39c93bac-ef3d-11e6-8c58-508eb76053b6.png)

Let's have a brief explanation for each component in the figure. Each circle represents a **unit** (or a **neuron**). And each square represents a calculation. The left most three units form the **input layer**. The neuron with an **h** inside is the only neuron the output layer of this neural network has. 

![The input to the output unit](https://cloud.githubusercontent.com/assets/14886380/22798974/04664f5c-ef3f-11e6-81cf-2b259680546c.png)

Recall that for a biological neuron, there exists a threshold for such a neuron to be activated. In our neural network, the neuron will calculate the input with an **activation function** and send the result as the output. One of the biggest advantages of the activation function is that the function could be **any** function. It means you can use step, polynomial or sigmoid functions of your choice to build your model. The output unit returns the result of **f(h)**, where **h** is the input to the output unit, and **y** is the output of the neural network.

if you let **f(h) = h** as your activation function, then the output of the network will be this, note that here **y = f(h)**.

![The output](https://cloud.githubusercontent.com/assets/14886380/22799516/fcada2f4-ef40-11e6-9570-f95a61ff0a9c.png)

You are correct if you think this is just the linear regression model. Once you start using activation functions that are **continuous** and **differentiable**, it's possible to train the network using **gradient descent**. We gonna need the first derivative of the activation function. Before we start, let's code the **dead simple neural network** in Python. For the activation function, we will use the sigmoid function.

![The sigmoid function](https://cloud.githubusercontent.com/assets/14886380/22743129/36eee4b6-ee54-11e6-8163-35c7c641d226.png)

```
import numpy as np

def sigmoid(x):
    # sigmoid function
    return 1/(1 + np.exp(-x))

inputs = np.array([0.7, -0.3])
weights = np.array([0.1, 0.8])
bias = -0.1

# calculate the output
output = sigmoid(np.dot(weights, inputs) + bias)

print('Output:')
print(output)
```

You can find the code [here](https://github.com/Kulbear/deep-learning-nano-foundation/tree/master/lectures).

## Refresher: The Gradient Descent
The gradient descent is one of the simplest methods we can apply to multiple algorithms for an optimization purpose.  

Frist of all, you probably realize that we need a metric to measure our prediction error. There are many metrics we can choose, such as **mean squared error (MSE)**, **mean absolute error (MAE)**, **R2-score** and so on. It would be another article if you want to dive into those concepts and have a clear understanding which one is best in a specific case. 

Google and Wikipedia are always your best friends. Beside from these resources, I'd like to share a post on [StackExchange](http://stats.stackexchange.com/a/48268) if you want a quick shot about how to select a better metric for your case from **MSE** and **MAE**.
